pkg/maps/lbmap/lbmap.go

```diff
@@ -147,24 +65,28 @@ func updateService(key ServiceKey, value ServiceValue) error {
 	return key.Map().Update(key.ToNetwork(), value.ToNetwork())
 }
 
-// DeleteService deletes a service from the lbmap. key should be the master (i.e., with backend set to zero).
+// DeleteService deletes a legacy service from the lbmap. The given key has to
+// be of the master service.
 func DeleteService(key ServiceKey) error {
 	mutex.Lock()
 	defer mutex.Unlock()
 
-	return deleteServiceLocked(key)
+	err := deleteServiceLocked(key)
+	if err != nil {
+		return err
+	}
+
+	cache.delete(key)
+
+	return nil
 }
 
 func deleteServiceLocked(key ServiceKey) error {
 	err := key.Map().Delete(key.ToNetwork())
 	if err != nil {
 		return err
 	}
-	err = lookupAndDeleteServiceWeights(key)
-	if err == nil {
-		cache.delete(key)
```

Comment:
This was removed from the common case, which means that the `UpdateService()` caller no longer removes the key from the cache when removing old backends that are no longer needed. Is this intentional?

Reply:
* `cache.delete(key)` removes the service entry from `lbmapCache.entries`.
* `deleteServiceLocked(key)` is called from `UpdateService` when removing legacy service entries. This can happen when a svc legacy is updated with an empty list of endpoints before its removal (such sequence of events has been observed on k8s).

So, this can lead to situations where `lbmapCache.removeServiceV2(..)` fails due to the svc cache entry not found due to ^^. We should only delete the cache entry upon a svc removal. I'm planning to backport this fix to v1.3/v1.4.